### PR TITLE
[Bugfix] [Numpy] Add `kAddTo` and kNullOp to Transpose

### PR DESCRIPTION
**Pull request by sxjscience**
## Description ##

- Add kAddTo and kNullOp to np.transpose and fix https://github.com/apache/incubator-mxnet/issues/16789
- Fix the problem that the previous np.tranpose support duplicated axis:

Previously, we can run the following code without any problem:
```python
import mxnet as mx
mx.npx.set_np()
a = mx.np.ones((3, 4, 5))
b = mx.np.transpose(a, axes=(0, 0, 1))
print(b.asnumpy())
```

Now, we will raise an MXNetError

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at https://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] kAddTo and kNullOp to transpose, test

### CODE REVIEW

**Comment by sxjscience** [comment]
@leezu @zburning This should fix the XLNet problem.

File: src/operator/tensor/matrix_op.cc [hunk 1]
```diff
@@ -283,11 +283,12 @@ static void TransposeComputeExCPU(const nnvm::NodeAttrs& attrs,
...
   }
   const TransposeParam& param = nnvm::get<TransposeParam>(attrs.parsed);
-  CHECK_EQ(req[0], kWriteTo) << "Transpose does not support kWriteInplace and kAddTo";
+  CHECK(req[0] == kWriteTo || req[0] == kAddTo) <<
+      "Transpose only supports kNullOp, kWriteTo and kAddTo";
   CHECK_EQ(inputs.size(), 1U);
   CHECK_EQ(outputs.size(), 1U);
 
-  if (SupportMKLDNNTranspose(param, inputs[0])) {
+  if (SupportMKLDNNTranspose(param, inputs[0]) && req[0] == kWriteTo) {
```
Comment (TaoLv) [inline src/operator/tensor/matrix_op.cc]: Thanks!

File: src/operator/tensor/pseudo2DTranspose_op-inl.cuh [hunk 2]
```diff
@@ -39,22 +39,29 @@ namespace mxnet {
 namespace op {
 namespace cuda {
 
-
-template <typename DType, typename CType>
+/*!
+ * \brief The `transpose_pseudo2D` based on chosen vectorized types. It transpose an array of
```
Comment (sxjscience) [inline src/operator/tensor/pseudo2DTranspose_op-inl.cuh]: @ptrendx I've also added the doc here for the `transpose_pseudo2D`. Correct me if you find any problem.

File: src/operator/tensor/pseudo2DTranspose_op-inl.cuh [hunk 3]
```diff
@@ -39,22 +39,29 @@ namespace mxnet {
...
+ * \tparam DType Data type
+ * \tparam CType The type to load the data.
+ * \tparam TSR the vectorized ratio.
+ * \tparam is_addto Whether to perform out += transpose(data) or out = transpose(data)
+ */
+template <typename DType, typename CType, int TSR, bool is_addto>
 __global__ void transpose_pseudo2D(DType* out, DType* inp,
                                    const index_t m, const index_t n,
                                    const index_t nIterY, const index_t nIterZ) {
-  const index_t TSR = sizeof(CType)/sizeof(DType);  // TypeSizeRatio
```
Comment (ptrendx) [inline src/operator/tensor/pseudo2DTranspose_op-inl.cuh]: Why did you move this to template argument? It should be fine here (you can even make it `constexpr`).

File: src/operator/tensor/pseudo2DTranspose_op-inl.cuh [hunk 4]
```diff
@@ -78,23 +85,34 @@ __global__ void transpose_pseudo2D(DType* out, DType* inp,
         }
         __syncthreads();
 
-        // read from shared to registers
-        transp_t tmp[TSR];
+        // read from shared to local registers
+        CType tmp[TSR];
         #pragma unroll
         for (index_t i = 0; i < TSR; i++) {
+          DType* tmp_dptr = reinterpret_cast<DType*>(&tmp[i]);
```
Comment (ptrendx) [inline src/operator/tensor/pseudo2DTranspose_op-inl.cuh]: Hmmm, I'm not sure if this will still be handled properly by the compiler, let me test that.

File: src/operator/tensor/pseudo2DTranspose_op-inl.cuh (hunk 3 above)
Comment (sxjscience) [reply src/operator/tensor/pseudo2DTranspose_op-inl.cuh]: I moved it to template to avoid the divide by zero error. Because in the other loop we are using switch to switch over the size of DType.

Comment (sxjscience) [reply src/operator/tensor/pseudo2DTranspose_op-inl.cuh]: Reports a `use too much shared memory error` after I change this. May be it's feasible to first put it into the template

Comment (sxjscience) [reply src/operator/tensor/pseudo2DTranspose_op-inl.cuh]: Okay, I think I know where the problem is. Let me try another time.

File: src/operator/tensor/pseudo2DTranspose_op-inl.cuh (hunk 4 above)
Comment (sxjscience) [reply src/operator/tensor/pseudo2DTranspose_op-inl.cuh]: Have you find any problem? I can compile that and it also passed CI.

File: src/operator/tensor/matrix_op-inl.h [hunk 5]
```diff
@@ -323,62 +336,77 @@ void TransposeImpl(RunContext ctx,
...
+      Tensor<xpu, 1, DType> in = src.get_with_shape<xpu, 1, DType>(mshadow::Shape1(src.Size()), s);
+      Tensor<xpu, 1, DType> out = ret.get_with_shape<xpu, 1, DType>(mshadow::Shape1(ret.Size()), s);
+      if (!is_addto) {
+        // Use memcpy to accelerate the speed
+        Copy(out, in, s);
+      } else {
+        mxnet_op::Kernel<mxnet_op::op_with_req<mshadow_op::identity, kAddTo>, xpu>::Launch(
+            s, ret.Size(), out.dptr_, in.dptr_);
+      }
+      return;
```
Comment (reminisce) [inline src/operator/tensor/matrix_op-inl.h]: `return` can be moved out of the macro to simplify expanded code a little bit.

File: src/operator/tensor/matrix_op-inl.h [hunk 6]
```diff
@@ -323,62 +336,77 @@ void TransposeImpl(RunContext ctx,
...
-        } else {
-          out = in.T();
-        }
+      Tensor<xpu, 2, DType> in = src.get<xpu, 2, DType>(s);
+      Tensor<xpu, 2, DType> out = ret.get<xpu, 2, DType>(s);
+      if (ctx.get_ctx().dev_mask() == cpu::kDevMask) {
+        Transpose2D<DType, is_addto>(in.dptr_, out.dptr_, in.shape_[0], in.shape_[1]);
       } else {
-        Copy(out, in, s);
+        LOG(FATAL) << "Not Implemented. We should never reach here. Report an issue in Github.";
```
Comment (reminisce) [inline src/operator/tensor/matrix_op-inl.h]: Can we improve the error message by pointing out the case should have been handled by `transpose_pseudo2D`?

File: src/operator/tensor/matrix_op-inl.h [hunk 7]
```diff
@@ -323,62 +336,77 @@ void TransposeImpl(RunContext ctx,
...
   if (isPseudo2DTranspose(axes)) {
     MSHADOW_TYPE_SWITCH(ret.type_flag_, DType, {
-      transpose_pseudo2D<DType>(ret, src, axes, s);
+      transpose_pseudo2D<DType, is_addto>(ret, src, axes, s);
     });
     return;
   }
 #endif
+  // Special handle the identity case
+  if (isIdentityTranspose(axes)) {
```
Comment (reminisce) [inline src/operator/tensor/matrix_op-inl.h]: `isIdentityTranspose` -> `IsIdentityTranspose`.

File: tests/python/unittest/test_numpy_op.py [hunk 8]
```diff
@@ -1313,45 +1315,56 @@ def __init__(self, axes=None):
 
         def hybrid_forward(self, F, a):
             return F.np.transpose(a, self.axes)
+    test_workloads = [[(), [(), None]],
```
Comment (reminisce) [inline tests/python/unittest/test_numpy_op.py]: Seems `ndim=1` is not tested?

File: src/operator/tensor/matrix_op-inl.h (hunk 5 above)
Comment (sxjscience) [reply src/operator/tensor/matrix_op-inl.h]: moved

File: src/operator/tensor/pseudo2DTranspose_op-inl.cuh (hunk 4 above)
Comment (ptrendx) [reply src/operator/tensor/pseudo2DTranspose_op-inl.cuh]: Hi, I only now had a chance to look into it. No, there is not any problem - I was worried that the compiler could get confused by this and put the `tmp` array in local memory instead of registers, but I tested and it does not do it.

Comment (sxjscience) [reply src/operator/tensor/pseudo2DTranspose_op-inl.cuh]: Great! Thanks for clarifying 👍

File: src/operator/tensor/matrix_op-inl.h [hunk 9]
```diff
@@ -323,62 +336,79 @@ void TransposeImpl(RunContext ctx,
...
-     }
      case 2: {
-      mshadow::Tensor<xpu, 2, DType> in = src.FlatTo2D<xpu, DType>(s);
-      mshadow::Tensor<xpu, 2, DType> out = ret.FlatTo2D<xpu, DType>(s);
-
-      if (axes[0] == 1 && axes[1] == 0) {
-        if (ctx.get_ctx().dev_mask() == cpu::kDevMask) {
-          Transpose2D<DType>(in.dptr_, out.dptr_, in.shape_[0], in.shape_[1]);
-        } else {
-          out = in.T();
```
[the remark this answers is not in the document]
Comment (sxjscience) [reply src/operator/tensor/matrix_op-inl.h]: @ChaiBapchya We have already had `transpose_pseudo2D`, which covers the case of 2D Transpose in GPU.

[the remark this answers is not in the document]
Comment (ChaiBapchya) [reply src/operator/tensor/matrix_op-inl.h]: Ah right..

**Comment by TaoLv** [comment]
@sxjscience Is it possible to create a case that Transpose has req kAddTo? I'm considering how to support that in MKL-DNN backend. Thanks!

**Comment by ptrendx** [comment]
@TaoLv just transpose the weight before using it (e.g. take nchw weight, transpose and use in nhwc convolution), and set grad_req as 'add' in simple_bind (gluon example would be pretty similar).
